hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -76,85 +77,22 @@ public SCMContainerPlacementRackScatter(final NodeManager nodeManager,
     this.metrics = metrics;
   }
 
-  /**
-   * Called by SCM to choose datanodes.
-   *
-   *
-   * @param excludedNodes - list of the datanodes to exclude.
-   * @param favoredNodes - list of nodes preferred. This is a hint to the
-   *                     allocator, whether the favored nodes will be used
-   *                     depends on whether the nodes meets the allocator's
-   *                     requirement.
-   * @param nodesRequiredToChoose - number of datanodes required.
-   * @param dataSizeRequired - size required for the container.
-   * @param metadataSizeRequired - size required for Ratis metadata.
-   * @return List of datanodes.
-   * @throws SCMException  SCMException
-   */
-  @Override
-  protected List<DatanodeDetails> chooseDatanodesInternal(
-      final List<DatanodeDetails> excludedNodes,
-      final List<DatanodeDetails> favoredNodes,
-      final int nodesRequiredToChoose, final long metadataSizeRequired,
-      final long dataSizeRequired) throws SCMException {
-    if (nodesRequiredToChoose <= 0) {
-      String errorMsg = "num of nodes required to choose should bigger" +
-          "than 0, but the given num is " + nodesRequiredToChoose;
-      throw new SCMException(errorMsg, null);
-    }
-    metrics.incrDatanodeRequestCount(nodesRequiredToChoose);
-    int nodesRequired = nodesRequiredToChoose;
-    int excludedNodesCount = excludedNodes == null ? 0 : excludedNodes.size();
-    List<Node> availableNodes = networkTopology.getNodes(
-        networkTopology.getMaxLevel());
-    int totalNodesCount = availableNodes.size();
-    if (excludedNodes != null) {
-      availableNodes.removeAll(excludedNodes);
+  public Set<DatanodeDetails> chooseNodesFromRacks(List<Node> racks,
```

Comment:
There is a lot of change in this class - is this method copied and moved from already existing code, or is the logic here new?

Reply:
It is same code and moved to a different function. Previously the function was very long.